src/aes128gcm.rs

```diff
@@ -23,174 +28,386 @@ const ECE_WEBPUSH_IKM_LENGTH: usize = 32;
             return Err(Error::InvalidRecordSize);
         }
+        let keyid = &input[ECE_AES128GCM_HEADER_LENGTH..ECE_AES128GCM_HEADER_LENGTH + keyid_len];
 
-        let salt = &payload[0..ECE_SALT_LENGTH];
-        if key_id_len != ECE_WEBPUSH_PUBLIC_KEY_LENGTH {
-            return Err(Error::InvalidKeyLength);
-        }
-        let key_id_pos = ECE_AES128GCM_HEADER_LENGTH;
-        let key_id = &payload[key_id_pos..key_id_pos + key_id_len];
-
-        let ciphertext_start = ECE_AES128GCM_HEADER_LENGTH + key_id_len;
-        if payload.len() == ciphertext_start {
-            return Err(Error::ZeroCiphertext);
-        }
-        let ciphertext = &payload[ciphertext_start..];
-        let cryptographer = crypto::holder::get_cryptographer();
-        let key = cryptographer.import_public_key(key_id)?;
-        Self::common_decrypt(local_prv_key, &*key, auth_secret, salt, rs, ciphertext)
+        Ok(Header { salt, rs, keyid })
     }
-}
 
-impl EceWebPush for Aes128GcmEceWebPush {
-    /// Always returns false because "aes128gcm" uses
-    /// a padding scheme that doesn't need a trailer.
-    fn needs_trailer(_: u32, _: usize) -> bool {
-        false
+    /// Write this `Header` at the start of the given output buffer.
+    ///
+    /// This assumes that the buffer has sufficient space for the data, and will
+    /// panic (via Rust's runtime safety checks) if it does not.
+    ///
+    pub fn write_into(&self, output: &mut [u8]) {
+        output[0..ECE_SALT_LENGTH].copy_from_slice(self.salt);
+        BigEndian::write_u32(&mut output[ECE_SALT_LENGTH..], self.rs);
+        output[ECE_AES128GCM_HEADER_LENGTH - 1] = self.keyid.len() as u8;
+        output[ECE_AES128GCM_HEADER_LENGTH..ECE_AES128GCM_HEADER_LENGTH + self.keyid.len()]
+            .copy_from_slice(self.keyid);
     }
 
-    fn allow_multiple_records() -> bool {
-        true
+    /// Get the size occupied by this header when written to the encrypted data.
+    ///
+    pub fn encoded_size(&self) -> usize {
+        ECE_AES128GCM_HEADER_LENGTH + self.keyid.len()
     }
+}
+
+/// Struct representing an individual plaintext record.
+///
+/// The encryption process splits up the input plaintext to fixed-size records,
+/// each of which is encrypted independently. This struct encapsulates all the
+/// data about a particular record. This diagram from the RFC may help you to
+/// visualize how this data gets encrypted:
+///
+///   +-----------+             content
+///   |   data    |             any length up to rs-17 octets
+///   +-----------+
+///        |
+///        v
+///   +-----------+-----+       add a delimiter octet (0x01 or 0x02)
+///   |   data    | pad |       then 0x00-valued octets to rs-16
+///   +-----------+-----+       (or less on the last record)
+///            |
+///            v
+///   +--------------------+    encrypt with AEAD_AES_128_GCM;
+///   |    ciphertext      |    final size is rs;
+///   +--------------------+    the last record can be smaller
```

Comment:
How does this look on rustdoc without ``` ?

Reply:
Ah, good reminder, thanks! (a high chance it'll look like crap without that)